File: web-app/src/services/taskService.ts
import { Task, TaskCreate, TaskUpdate } from '@/types/task'

const API_BASE_URL = 'http://localhost:8000/api'

class TaskService {
  async getAllTasks(): Promise<Task[]> {
    const response = await fetch(`${API_BASE_URL}/tasks`)
    if (!response.ok) {
      throw new Error('Failed to fetch tasks')
    }
    return response.json()
  }

  async createTask(taskData: TaskCreate): Promise<Task> {
    const response = await fetch(`${API_BASE_URL}/tasks`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(taskData),
    })
    if (!response.ok) {
      throw new Error('Failed to create task')
    }
    return response.json()
  }

  async updateTask(id: string, taskData: TaskUpdate): Promise<Task> {
    const response = await fetch(`${API_BASE_URL}/tasks/${id}`, {
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(taskData),
    })
    if (!response.ok) {
      throw new Error('Failed to update task')
    }
    return response.json()
  }

  async completeTask(id: string, completed: boolean): Promise<Task> {
    const response = await fetch(`${API_BASE_URL}/tasks/${id}/complete`, {
      method: 'PATCH',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ completed }),
    })
    if (!response.ok) {
      throw new Error('Failed to update task completion status')
    }
    return response.json()
  }

  async deleteTask(id: string): Promise<void> {
    const response = await fetch(`${API_BASE_URL}/tasks/${id}`, {
      method: 'DELETE',
    })
    if (!response.ok) {
      throw new Error('Failed to delete task')
    }
  }
}

export const taskService = new TaskService()

File: backend_workaround/services/task_service.py
from typing import List, Optional
from sqlmodel import Session
from uuid import UUID
from backend_workaround.models.task import Task, TaskCreate, TaskUpdate
from backend_workaround.repositories.task_repository import TaskRepository


class TaskService:
    def __init__(self, repository: TaskRepository):
        self.repository = repository

    def create_task(self, session: Session, task: TaskCreate) -> Task:
        return self.repository.create_task(session, task)

    def get_task(self, session: Session, task_id: UUID) -> Optional[Task]:
        return self.repository.get_task(session, task_id)

    def get_all_tasks(self, session: Session) -> List[Task]:
        return self.repository.get_tasks(session)

    def update_task(self, session: Session, task_id: UUID, task_update: TaskUpdate) -> Optional[Task]:
        task_data = task_update.model_dump(exclude_unset=True)
        return self.repository.update_task(session, task_id, task_data)

    def delete_task(self, session: Session, task_id: UUID) -> bool:
        return self.repository.delete_task(session, task_id)

    def complete_task(self, session: Session, task_id: UUID, completed: bool) -> Optional[Task]:
        return self.repository.update_task(session, task_id, {"completed": completed})